Remove requirement to keep TOAs grouped

dlakaplan opened the pull request:
TOAs no longer need to be grouped, although they still can be.  Slight change in order of operations/interpolation in `apply_clock_correction` that results in ~3e-15s changes.

(3rd attempt at this, after messing up previous PRs)

dlakaplan commented:
Addresses #1252

aarchiba commented on src/pint/toa.py:
```diff
@@ -306,6 +304,7 @@ def get_TOAs(
     if usepickle and updatepickle:
         log.info("Pickling TOAs.")
         save_pickle(t, picklefilename=picklefilename)
+    t.table = t.table.group_by("obs")
```
Why group the table rather than preserving the user's input order?

aarchiba commented on src/pint/simulation.py:
```diff
@@ -168,6 +168,7 @@ def make_fake_toas(ts, model, add_noise=False, name="fake"):
     for f in tsim.table["flags"]:
         f["name"] = name
 
+    tsim.table = tsim.table.group_by("obs")
```
Why group the table now it is no longer necessary?

aarchiba replied:
Why group the table now it is no longer necessary?

aarchiba commented on src/pint/toa.py:
```diff
@@ -419,6 +418,7 @@ def get_TOAs_list(
         t.compute_TDBs(method=tdb_method, ephem=ephem)
     if "ssb_obs_pos" not in t.table.colnames:
         t.compute_posvels(ephem, planets)
+    t.table = t.table.group_by("obs")
```
Why group the table?

aarchiba replied:
Why group the table?

aarchiba commented on src/pint/toa.py:
```diff
@@ -1130,18 +1130,14 @@ class TOAs:
-    access the contained information but the data may not be in the order you
-    expect: internally it is grouped by observatory (sorted by the observatory
-    object). Not all columns of the table are computed automatically, as their
+    The contents are stored in an :class:`astropy.table.Table`. Not all columns of the table are computed automatically, as their
```
It would be good to reflow this.

aarchiba commented on src/pint/toa.py:
```diff
@@ -1154,6 +1150,11 @@ class TOAs:
 
     >>> t['high', t['freq'] > 1*u.GHz] = "1"
 
+    TOAs are grouped by observatory by default, but this is just a convention.  To iterate over all observatory groups:
```
I think TOAs should be left in the order they were in the tim files.

aarchiba commented on tests/test_toa.py:
```diff
@@ -87,19 +87,17 @@ def test_mulit_obs(self):
-        assert toas.table["mjd"][1] == t3.mjd
-        assert toas.table["obs"][2] == site1.name
-        assert toas.table["mjd"][2] == t1.mjd
+        assert toas.table["obs"][0] == site1.name
```
What property is this test verifying? That the order of the result is just the concatenation of the originals? I don't understand what makes this the right behaviour.

aarchiba replied:
Can you add some tests that check what order the TOAs returned from various operations are in? I think we should default to preserving the order in the tim file unless the user explicitly re-orders the file, and I think the tests should verify this.

aarchiba replied:
Can you add some tests that check what order the TOAs returned from various operations are in? I think we should default to preserving the order in the tim file unless the user explicitly re-orders the file, and I think the tests should verify this.

aarchiba commented on tests/test_toa.py:
```diff
@@ -87,19 +87,17 @@ def test_mulit_obs(self):
-        assert toas.table["mjd"][1] == t3.mjd
-        assert toas.table["obs"][2] == site1.name
-        assert toas.table["mjd"][2] == t1.mjd
+        assert toas.table["obs"][0] == site1.name
```
What property is this test verifying? That the order of the result is just the concatenation of the originals? I don't understand what makes this the right behaviour.

aarchiba commented:
I think there is a discussion to be had: what order should TOAs be in, inside PINT? I suggest they stay in the order they were in `.tim` files unless users explicitly reorder them. Whatever the decision, we should have tests that verify that the order we claim TOAs will be in is the order they are actually in.

dlakaplan commented:
@aarchiba : some quick answers (can discuss further):
* Why group the table?  In some cases it could be useful, and I was trying to minimally change previous behavior.  If we discuss and decide that this isn't useful it's easy to now eliminate.  Or make it an option when loading TOAs.
* Seemingly trivial tests: most of these were more meaningful when grouping was more of a thing.  Again, I tried to just do minimal changes in case there was some other more subtle purpose that I didn't have time to figure out.  If we decide that they are deprecated we can change.

aarchiba commented:
Sorry for the doubled review comments, I don't know what happened with that.

dlakaplan commented:
@aarchiba / @scottransom : I am testing the result of sorting the TOAs.  To a large extent it seems fine, but when I compare the chi^2 they are not identical (differ at ~1e-14).  Some changes went away when I forced the first TOA to be the same, so it could be an absolute phase question, but not all.  Any reasons why the order should matter?

dlakaplan commented:
I added `tests/test_timing_model.py::test_resorting_toas` to test this.  There is one equality and one `isclose`, but I'd like to understand the reason that they are both not equal

aarchiba commented:
> @aarchiba / @scottransom : I am testing the result of sorting the TOAs. To a large extent it seems fine, but when I compare the chi^2 they are not identical (differ at ~1e-14). Some changes went away when I forced the first TOA to be the same, so it could be an absolute phase question, but not all. Any reasons why the order should matter?

The order definitely should not matter. Can you compare the residuals? Use argsort and compare residuals then reshuffle to reshuffle then residuals? Does `abs_phase=True` change anything?

dlakaplan commented on tests/test_timing_model.py:
```diff
@@ -388,6 +390,33 @@ def test_parfile_and_timfile_jumps(timfile_jumps):
+    i[[0, i0]] = i[[i0, 0]]
+    tcopy.table = tcopy.table[i]
+    rsort = pint.residuals.Residuals(tcopy, m)
+    assert r.calc_chi2() == rsort.calc_chi2()
```
```
tcopy = copy.deepcopy(t)
# sort by mjd
i = np.argsort(t.table["mjd_float"])
tcopy.table = tcopy.table[i]
rsort = pint.residuals.Residuals(tcopy, m)
print((rsort.time_resids - r.time_resids[i]))
```
gives:
```
[-3.33413344e-19 -3.33519223e-19 -3.33413344e-19 -3.33519223e-19
 -3.33413344e-19 -3.33519223e-19 -3.33413344e-19 -3.33519223e-19
 -3.33519223e-19 -3.33413344e-19 -3.33519223e-19 -3.33413344e-19
 -3.33466283e-19 -3.33466283e-19 -3.33519223e-19 -3.33466283e-19
 -3.33466283e-19 -3.33466283e-19 -3.33519223e-19 -3.33492753e-19
 -3.33492753e-19 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19 -3.33492753e-19 -3.33492753e-19
 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19 -3.33413344e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19 -3.33492753e-19 -3.33492753e-19
 -3.33492753e-19 -3.33492753e-19 -3.33492753e-19 -3.33492753e-19
 -3.33492753e-19 -3.33492753e-19 -3.33492753e-19 -3.33492753e-19
 -3.33492753e-19 -3.33492753e-19 -3.33519223e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19 -3.33519223e-19
 -3.33519223e-19 -3.33519223e-19] s
```

However, if I use `subtract_mean=False` in computing the residuals it gives 0.

aarchiba replied:
Looks like the mean subtraction is subtly order-dependent. To be fair, floating-point addition is subtly order-dependent, so that may be the cause.

dlakaplan commented on tests/test_timing_model.py:
```diff
@@ -388,6 +390,33 @@ def test_parfile_and_timfile_jumps(timfile_jumps):
+    rsort = pint.residuals.Residuals(tcopy, m)
+    assert r.calc_chi2() == rsort.calc_chi2()
+    # sort by MJD
+    tcopy = deepcopy(t)
```
In fact, even when the residuals agree the chi^2 does not.  I can compare `residuals / uncertainties` for both Residuals objects and those agree, but chi^2 does not.  Still digging...

aarchiba replied:
That sounds like there's some questionable mean(?) subtraction in the chi-squared calculation.

dlakaplan replied:
I don't think so.  I think it's just
```
return ((self.time_resids / toa_errors.to(u.s)) ** 2.0).sum().value
```

dlakaplan commented:
I tried this 100 times with random shuffles of 10 TOAs.  I see that the chi^2 values sometimes agree, sometimes disagree by +/- 5e-14.  So I suspect that this may be a floating point issue?

aarchiba commented on tests/test_toa_indexing.py:
```diff
@@ -85,7 +77,6 @@ def test_getitem_slice(c):
     s = toas[c]
     assert set(s.get_mjds()) == set(m[c])
     if len(s) > 0:
-        assert np.all(s.table["mjd_float"] == s.table.group_by("obs")["mjd_float"])
```
This test can be greatly simplified now that we preserve order - you can just check that `toas[c].get_mjds() == toas.get_mjds()[c]` and drop the whole rest of the testing code. Other indexing tests can be similarly modified to be both simpler and stricter.

aarchiba commented on tests/test_toa_indexing.py:
```diff
@@ -27,9 +27,6 @@ def test_get_TOAs(high_precision):
     assert isinstance(m, np.ndarray)
-    assert not np.all(
-        np.diff(m) > 0
-    ), "returned values should be grouped by observatory"
```
This should become a test that indeed all returned values are increasing (assuming that's what's in the input par).

aarchiba commented on tests/test_toa_indexing.py:
```diff
@@ -57,7 +51,6 @@ def test_getitem_boolean(c):
     assert len(s) == np.sum(c)
     assert np.all(s.get_mjds() == m[c])
     if len(s) > 0:
-        assert np.all(s.table["mjd_float"] == s.table.group_by("obs")["mjd_float"])
```
I think this test differs from the previous in that it checks that goruping is preserved, so it can go.

aarchiba commented on tests/test_timing_model.py:
```diff
@@ -385,6 +387,56 @@ def test_parfile_and_timfile_jumps(timfile_jumps):
+    r = pint.residuals.Residuals(t, m, subtract_mean=False)
+    tcopy = deepcopy(t)
+    i = np.arange(len(t))
+    np.random.shuffle(i)
```
Random shuffles are liable to cause a problem with tests that fail once and then not the next time. You can use `hypothesis`: `@given(permutations(np.arange(n)))`. This explores the space of permutations and finds the simplest one that fails the test.

dlakaplan replied:
ok, good to know.  but I can't quite figure out the syntax to define n properly using a fixture or class-based method.  I just pushed up revised tests done in a single class but I had to define n using a global.  is there a better way?  when I tried to define it as a fixture there were various syntax errors.

aarchiba replied:
The magic sauce is that `n` is supposed to be the length of the TOAs object, so you need to make or pick the TOAs object within hypothesis. I think `flatmap` is the right way to do this: https://hypothesis.readthedocs.io/en/latest/data.html#chaining-strategies-together bit I'd find it easier to use `@composite`:

```python
@composite
def toas_and_order(draw):
    toas_name = draw(select_from(["file1.tim", "file2.tim"]))
    toas = get_TOAs(toas_name)
    n = len(toas)
    ix = draw(permutations(np.arange(n)))
    return toas, ix
```
then you'd use `@given(toas_and_order())` to get that tuple inside your function.

I think there are examples of `@composite` doing this sort of thing elsewhere in our test suite.

dlakaplan replied:
That is handy.  I reworked it a bit so that it loads the data initially through a data class (unofficial) to speed things up, but I think this now does what it should

dlakaplan commented:
@scottransom : I verified that I can delete observations correctly from the Ter5 data you sent.  that seemed to be a place where grouping was an issue. I also changed "group" -> "cluster" the documentation and code about the "v" key, and put in a default `get_cluster()` call to make that work.  I removed other `pintk` group discussion.  Are there other aspects to test?

dlakaplan commented:
I'm not sure how to test deleting stashed TOAs, since they aren't visible.  how does that work?

scottransom commented:
I was just talking about the "deleting" that happens in the TOA table that the user sees when they stash.  That is one of the only places where a group_by() is explicitly used in the pintk code:

```
                # Delete the selected points
                self.psr.stashed = copy.deepcopy(self.psr.all_toas)
                self.psr.all_toas.table = self.psr.all_toas.table[
                    ~self.selected
                ].group_by("obs")
```

And your code already gets rid of that `group_by`, so if stashing/unstashing works, you should be good.

dlakaplan commented:
OK, yes.  Can stash/unstash.

aarchiba commented on tests/test_timing_model.py:
```diff
@@ -385,6 +394,60 @@ def test_parfile_and_timfile_jumps(timfile_jumps):
 
+class TOAOrderSetup:
+    parfile = os.path.join(datadir, "NGC6440E.par")
+    timfile = os.path.join(datadir, "NGC6440E.tim")
```
Is this a single-telescope file? I think it's just GBT data? It should be fine but it would be nice to have a mixed-telescope data set to ensure we catch weirdness with treating things differently by telescope.

dlakaplan replied:
I agree in principle, but the existing test data don't have lots of multi-telescope data and those that are there (1713) are very large.  I could possibly put in a new one.

dlakaplan replied:
OK, have a new test `test_toa_shuffle.py` which does the same routines but with multi-telescope multi-frequency simulated data.

aarchiba commented on tests/test_timing_model.py:
```diff
@@ -42,6 +47,9 @@ def timfile_nojumps():
     return get_TOAs(os.path.join(datadir, "NGC6440E.tim"))
 
 
+len_timfile_nojumps = len(get_TOAs(os.path.join(datadir, "NGC6440E.tim")))
```
Can we avoid reloading this file, or save time?

aarchiba commented on tests/test_toa_shuffle.py:
```diff
@@ -0,0 +1,88 @@
+        t.write_TOA_file(f)
+    f.seek(0)
+    t = toa.get_TOAs(f)
+    r = pint.residuals.Residuals(t, model, subtract_mean=False)
```
Is this ever used?

aarchiba replied:
Oh, I see that you can access this directly from anywhere.

dlakaplan commented on tests/test_toa_shuffle.py:
```diff
@@ -0,0 +1,88 @@
+    for t in fakes:
+        t.write_TOA_file(f)
+    f.seek(0)
+    t = toa.get_TOAs(f)
```
I didn't directly do this test, but I did extensively test the clock corrections in the old version and without the grouping.  Prior to a the reordering discussed above they were identical.  Now they are very close (just due to when the clock data are interpolated).  But I think I could put in another test

dlakaplan replied:
that is now done.

dlakaplan commented:
I added a test to shuffle the lines of a tim file before loading (perhaps in a clumsy way) and then verify the clock corrections compared to the non-shuffled version.

aarchiba commented on tests/test_toa_shuffle.py:
```diff
@@ -0,0 +1,124 @@
+        [
+            x
+            for x in lines[1:]
+            if not (x.startswith("C") or x.startswith("#") or len(x.strip()) == 0)
```
Why do this? The shuffle should still work, shouldn't it?

aarchiba replied:
Ah, I see. Makes the TOA permutation the same as the line permutation.

aarchiba commented on tests/datafile/shuffletest.tim:
```diff
@@ -0,0 +1,123 @@
+FORMAT 1
+#J0523-7125_PSMC.rf.f4T 1382.000000 55415.829511131335920 4407.496  pks 
```
Is there any reason this can't be used as an example file with multiple telescopes for other test code? The name suggests a more specific nature than we necessarily need. If it's really only good for the one use, maybe trim it to three or so TOAs from each telescope and put it in a triple-quoted string?

dlakaplan replied:
It's from an unpublished project, and I didn't want to include the par file (which might have been helpful for other tests).  Ultimately for this test the TOAs don't really matter so I could just make up numbers and put it as a string - I was just exploring the testing possibilities.

dlakaplan replied:
I changed it to fake data.

dlakaplan commented:
@aarchiba @scottransom : do you think this can be merged?

scottransom commented:
I certainly do.  Would like to hear from Anne first, though.

aarchiba commented:
Looks good to me.
